File: src/Java/myservicebus-amazon-sqs/src/main/java/com/myservicebus/amazon/sqs/AmazonSqsMessageMapper.java
package com.myservicebus.amazon.sqs;

import com.myservicebus.MessageHeaders;
import com.myservicebus.serialization.MassTransitHeaderConvention;
import software.amazon.awssdk.services.sns.model.PublishRequest;
import software.amazon.awssdk.services.sqs.model.Message;
import software.amazon.awssdk.services.sqs.model.SendMessageRequest;

import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

final class AmazonSqsMessageMapper {
    private static final String CONTENT_TYPE = "Content-Type";

    private AmazonSqsMessageMapper() { }

    static SendMessageRequest sqsRequest(String queueUrl, byte[] body, String contentType) {
        return SendMessageRequest.builder().queueUrl(queueUrl)
                .messageBody(new String(body, StandardCharsets.UTF_8))
                .messageAttributes(Map.of(CONTENT_TYPE,
                        software.amazon.awssdk.services.sqs.model.MessageAttributeValue.builder()
                                .dataType("String").stringValue(contentType).build()))
                .build();
    }

    static PublishRequest snsRequest(String topicArn, byte[] body, String contentType) {
        return PublishRequest.builder().topicArn(topicArn)
                .message(new String(body, StandardCharsets.UTF_8))
                .messageAttributes(Map.of(CONTENT_TYPE,
                        software.amazon.awssdk.services.sns.model.MessageAttributeValue.builder()
                                .dataType("String").stringValue(contentType).build()))
                .build();
    }

    static Map<String, Object> headers(Message message, String faultAddress) {
        Map<String, Object> headers = new HashMap<>();
        var contentType = message.messageAttributes().get(CONTENT_TYPE);
        headers.put(MassTransitHeaderConvention.INSTANCE.getContentTypeHeader(),
                contentType != null ? contentType.stringValue() : "application/vnd.masstransit+json");
        if (message.messageId() != null) headers.put("message_id", message.messageId());
        String receiveCount = message.attributesAsStrings().get("ApproximateReceiveCount");
        if (receiveCount != null) headers.put(MessageHeaders.REDELIVERY_COUNT,
                Math.max(0, Integer.parseInt(receiveCount) - 1));
        if (faultAddress != null) headers.putIfAbsent(
                MassTransitHeaderConvention.INSTANCE.getFaultAddressHeader(), faultAddress);
        return headers;
    }
}

File: src/MyServiceBus.AmazonSqs/AmazonSqsMessageMapper.cs
using Amazon.SimpleNotificationService.Model;
using Amazon.SQS.Model;
using MyServiceBus.Serialization;
using SnsMessageAttributeValue = Amazon.SimpleNotificationService.Model.MessageAttributeValue;
using SqsMessageAttributeValue = Amazon.SQS.Model.MessageAttributeValue;

namespace MyServiceBus;

internal static class AmazonSqsMessageMapper
{
    private const string ContentTypeAttribute = "Content-Type";
    private static readonly IMessageHeaderConvention HeaderConvention = MassTransitHeaderConvention.Instance;

    public static SendMessageRequest CreateSqsRequest(string queueUrl, byte[] body, string contentType) =>
        new()
        {
            QueueUrl = queueUrl,
            MessageBody = System.Text.Encoding.UTF8.GetString(body),
            MessageAttributes = new Dictionary<string, SqsMessageAttributeValue>
            {
                [ContentTypeAttribute] = new() { DataType = "String", StringValue = contentType }
            }
        };

    public static PublishRequest CreateSnsRequest(string topicArn, byte[] body, string contentType) =>
        new()
        {
            TopicArn = topicArn,
            Message = System.Text.Encoding.UTF8.GetString(body),
            MessageAttributes = new Dictionary<string, SnsMessageAttributeValue>
            {
                [ContentTypeAttribute] = new() { DataType = "String", StringValue = contentType }
            }
        };

    public static Dictionary<string, object> CreateHeaders(Message message, Uri? faultAddress)
    {
        var headers = new Dictionary<string, object>(StringComparer.Ordinal);
        if (message.MessageAttributes.TryGetValue(ContentTypeAttribute, out var contentType) &&
            !string.IsNullOrWhiteSpace(contentType.StringValue))
            headers[HeaderConvention.ContentTypeHeader] = contentType.StringValue;
        else
            headers[HeaderConvention.ContentTypeHeader] = InboundMessageResolver.EnvelopeContentType;

        if (!string.IsNullOrWhiteSpace(message.MessageId))
            headers["message_id"] = message.MessageId;
        if (message.Attributes.TryGetValue("ApproximateReceiveCount", out var count))
            headers[MessageHeaders.RedeliveryCount] = Math.Max(0, int.Parse(count, System.Globalization.CultureInfo.InvariantCulture) - 1);
        if (faultAddress is not null && !headers.ContainsKey(HeaderConvention.FaultAddressHeader))
            headers[HeaderConvention.FaultAddressHeader] = faultAddress.ToString();
        return headers;
    }
}
